Add Danfoss Ally thermostat and derivatives to ZHA

Caius-Bonus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds support for the proprietary attributes of the Danfoss Ally thermostat and derivatives.

It does this in 5 parts:
1. It implements: the (binary) sensor, select, switch and number entities.
    a. It implements certain sensors using a more appropriate type than used in the zha-quirks (if changed in the zha-quirks, no changes are required in homeassistant core though).
2. It adds english strings for the entities.
3. It adds the BitMapSensor Class to be used for bitmap sensors.
4. It adds tests for the BitMap Sensor
5. It adds manufacturer specific cluster handlers for the Danfoss diagnostic, userinterface and thermostat clusters.




This makes it possible to fully control the Thermostatic Radiator Valves just like with Zigbee2MQTT.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
https://github.com/zigpy/zha-device-handlers/issues/1726 (It will now have this)
https://github.com/zigpy/zha-device-handlers/issues/717  (It will now have this)
https://github.com/zigpy/zha-device-handlers/issues/2728 (It will now have this)
https://github.com/home-assistant/core/issues/106557 (both issues will be gone)
- These PRs are related to this issue: https://github.com/zigpy/zha-device-handlers/issues/2115 (I don't know what could cause this, but it might change)
- This PR is dependent on PRs: 
  * https://github.com/home-assistant/core/pull/101725
  * https://github.com/home-assistant/core/pull/101709
  * https://github.com/home-assistant/core/pull/102489 (use same method of referencing quirk ids)
  * https://github.com/zigpy/zha-device-handlers/pull/2150
  * https://github.com/home-assistant/core/pull/116071

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Caius-Bonus committed:
zha integration: Add danfoss specific clusters and attributes; add thermostat.pi_heating_demand and thermostat_ui.keypad_lockout

Caius-Bonus committed:
zha integration: fix Danfoss thermostat viewing direction not working because of use of bitmap8 instead of enum8

Caius-Bonus committed:
ZHA Integration: add missing ThermostatChannelSensor

Caius-Bonus committed:
ZHA integration: format using black

Caius-Bonus committed:
zha integration: fix flake8 issues

Caius-Bonus committed:
Merge branch 'home-assistant:dev' into dev

Caius-Bonus commented:
I'll turn this into a draft for now, because I still need to improve some things both here and in zha-device-handlers.

Caius-Bonus committed:
ZHA danfoss: Add MinHeatSetpointLimit, MaxHeatSetpointLimit, add reporting and read config for danfoss and keypad_lockout.

Caius-Bonus committed:
Merge branch 'home-assistant:dev' into dev

Caius-Bonus committed:
ZHA danfoss: fix mypy complaining about type of _attr_entity_category

Caius-Bonus committed:
ZHA danfoss: ruff fix

Caius-Bonus commented:
Okay, Issues are resolved.

Caius-Bonus commented:
No reason to close mr. robot.

Caius-Bonus commented:
Still no reason to close.

edenhaus reviewed (dismissed):
Can you please rebase and fix the merge conflicts?

After that, I can start reviewing your PR

Caius-Bonus committed:
Merge branch 'dev' of https://github.com/home-assistant/core into home-assistant-dev

Caius-Bonus committed:
Merge branch 'home-assistant-dev' into dev

Caius-Bonus committed:
fix tests

Caius-Bonus commented:
I rebased the code and fixed the CLA issue. A lot has changed and I will have to retest the code with the physical device. Also, the merge is dependent on the quirk in zha-device-handlers, which isn't merged yet either, but I don't foresee much changing. Thanks for taking a look.

Caius-Bonus committed:
pylint: disable-next=hass-invalid-inheritance

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
fix pylint tests

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
Merge branch 'home-assistant:dev' into dev

Caius-Bonus committed:
refactoring

Caius-Bonus committed:
remove scheduled setpoint

Caius-Bonus committed:
remove scheduled setpoint in manufacturer specific

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus commented:
I retested and it works now. Ready for review

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
refactor

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
change cluster ids

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
remove custom clusters

Caius-Bonus committed:
code quality

Caius-Bonus committed:
Merge pull request #2 from Caius-Bonus/RemoveCustomClusters

Remove custom clusters

Caius-Bonus committed:
Merge branch 'dev' into dev

TheJulianJES commented on homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py:
```diff
@@ -410,26 +425,36 @@ class DanfossTRVChannel(ClusterHandler):
+    @classmethod
+    def matches(cls, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> bool:
+        """Filter the cluster match for specific devices."""
+        return cluster.endpoint.device.manufacturer in DANFOSS_TRVS
```
You might be able to use `cluster.endpoint.device.quirk_class` here.
However, https://github.com/home-assistant/core/pull/93268/files changed the "matches" to only check the "last two strings" when separated by a dot (`.`). So the code here should also check like that.

IMO, this "last two string getter" should be put in the base class though, so we don't do the same thing in multiple places.
Not sure if it should be done in this PR though.

---

But tbh, I don't like this whole "quirk class" matching system, as it can change during refactors and with now only taking "the last two strings", it's not even required that the manufacturer appears in that "quirk class" depending on how zha-quirks is structured for the manufacturer.

For example, I'd rather prefer something like a "quirk ID" (just a simple string that's set in each quirk that needs to be matched against in ZHA) that uniquely identifies a quirk — not something that can change like a file name/when refactoring). (But this needs to be discussed first and not done here)

Caius-Bonus replied:
I agree that there should be a more maintainable way to match quirks. Although I do believe that the method now is better than my first attempt :). I'll do testing on device again to validate tomorrow.

Caius-Bonus committed:
match clusters in manufacturerspecific on quirk class

Caius-Bonus committed:
fix comment

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
fix match on quirk in manufacturerspecific.py

Caius-Bonus committed:
correctly extend cluster handlers in manufacturerspecific.py and remove workaround for illegal use of attribute updated signals in climate.py

Caius-Bonus committed:
fix style

Caius-Bonus committed:
allow non-danfoss thermostats to work in manufacturerspecific.py

Caius-Bonus committed:
correct order of init of parent and subclasses in manufacturerspecific.py

Caius-Bonus committed:
improve entity names

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus committed:
fix pylint

Caius-Bonus commented:
Tested and fully working again

Caius-Bonus committed:
explicitly state changing size of tuple

Caius-Bonus committed:
ignore tuple size change error

Caius-Bonus committed:
really ignore error

Caius-Bonus committed:
Merge branch 'dev' into dev

Caius-Bonus commented:
Looking at my code, I don't really like the way the ClusterHandlers are implemented. Because I need to have a ClusterHandler for a zigbee standard cluster with additional attributes, every instance of ThermostatClusterHandler has to be an instance of ManufacturerSpecificThermostat. It would be better to match based on the quirk (using an ID as TheJulianJES proposed) for these specific ClusterHandlers. Would you be open to me opening a pull-request for this? This would also be a way to handle the Xiaomi doorlock cluster (removing the "ugly hack", as it is called in the code).

Caius-Bonus committed:
initial

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
match on specific name and quirk name

Caius-Bonus committed:
don't restructure file as it is out of scope

Caius-Bonus committed:
move back

Caius-Bonus committed:
remove unnecessary change

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
Merge branch 'home-assistant:dev' into dev

Caius-Bonus committed:
remove code duplication

Caius-Bonus committed:
reduce code duplication

Caius-Bonus committed:
empty line

Caius-Bonus committed:
remove unused variable

Caius-Bonus committed:
end file on newline

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
comply with recent PRs

Caius-Bonus committed:
correctly initialize all attributes

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
comply with recent PRs

Caius-Bonus committed:
make class variables private

Caius-Bonus committed:
forgot one reference

Caius-Bonus committed:
swap 2 lines for consistency

Caius-Bonus committed:
reorder 2 lines

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
align with recent PR

Caius-Bonus committed:
store cluster handlers in only one place

Caius-Bonus committed:
edit tests

Caius-Bonus committed:
Merge branch 'dev' into MatchCustomCluster

Caius-Bonus committed:
use correct device for quirk id

Caius-Bonus committed:
change quirk id

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
even if there is a quirk id, it doesn't have to have a specific cluster handler

Caius-Bonus committed:
add tests

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
Merge branch 'AddDanfossEntities' into MatchCustomCluster

Caius-Bonus committed:
Merge pull request #3 from Caius-Bonus/MatchCustomCluster

Match custom cluster

Caius-Bonus committed:
use quirk id for manufacturer specific cluster handlers

Caius-Bonus committed:
use quirk_ids instead of quirks_classes

Caius-Bonus committed:
rename quirk_id

Caius-Bonus committed:
rename quirk_id

Caius-Bonus committed:
forgot to rename here

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
rename id

Caius-Bonus committed:
Merge branch 'dev' into MatchCustomCluster

Caius-Bonus committed:
add tests

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
fix tests

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
use quirk ids from zha_quirks

Caius-Bonus committed:
use quirk id from zha_quirks

Caius-Bonus commented:
It now refers to quirk ids defined in zha_quirks.

Caius-Bonus committed:
Merge branch 'dev' into MatchCustomCluster

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
merge

Caius-Bonus committed:
wrong translation

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

AllardKatan commented:
Hi, I am eagerly awaiting the improved ZHA support for my new Popp trv's that this PR will bring. Is there anything I can do to help? I wouldn't call myself an experienced python coder but if it's testing or something basic that just needs time I'm willing to contribute.

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus commented:
@AllardKatan , Unfortunately it isn't trivial to test this for yourself. Waiting until it is merged would be my advice.

Caius-Bonus committed:
sync changes with ZCL branch

Caius-Bonus committed:
sync

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
style

Caius-Bonus committed:
merge error

Caius-Bonus committed:
move bitmapSensor

Caius-Bonus committed:
merge error

Caius-Bonus committed:
merge error

Caius-Bonus committed:
watch the capitals

Caius-Bonus committed:
fix entity categories

Caius-Bonus commented:
Seems to work fine, except for the dependency on zhaquirks. I think 2024.3.0 is a realistic target.

TheJulianJES commented:
Let's keep it in the draft until the quirks PR is merged. That way there's no confusion about when it can be merged.
Of course, this PR can already be reviewed, as it'll (mostly) stay as-is.

TheJulianJES commented on homeassistant/components/zha/sensor.py:
```diff
@@ -1367,3 +1368,164 @@ class AqaraCurtainHookStateSensor(EnumSensor):
+    _attr_translation_key: str = "adaptation_run_status"
+    _attr_entity_category = EntityCategory.DIAGNOSTIC
+    _default_value = "Nothing"
+    _bitmap = {
+        "In Progress": 0x0001,
+        "Run Successful": 0x0002,
+        "Valve characteristic lost": 0x0004,
+    }
```
Hmm, it's not ideal that these aren't translated.
Also, the capitalization needs to be changed to match the HA standard (for all BitMapSensors): e.g. "In progress", ..

Caius-Bonus replied:
Thanks, missed those capitals. Wasn't there a PR that mentioned the translation of enums (https://github.com/home-assistant/core/pull/96553 was closed)? I take it that that hasn't been done yet. Maybe I could take a look at that. The translation of enums could be done in a separate PR and the translation of bitmaps here as it is (for now) only needed for Danfoss.

Caius-Bonus replied:
The extra_state_attributes can be easily translated for the bitmap, but the state of the entity doesn't seem to work for some reason. But if I can get that to work, translation of enums is also going to be easy.

Caius-Bonus replied:
Okay, looked at the code for a couple of hours to realize the translations didn't get regenerated. It worked after all. I added every enum in strings.json. However for enums to work, new merge requests are required. I could also manually change the camelcase to snake case in my own enums, but that would make them inconsistent with zigpy.

Caius-Bonus committed:
more decapitalization

Caius-Bonus committed:
translate BitmapSensor

Caius-Bonus committed:
translate all enums

Caius-Bonus committed:
translate all enums

Caius-Bonus committed:
don't convert camelcase to snakecase

Caius-Bonus committed:
don't change enums at all

Caius-Bonus committed:
remove comments

peckpeck commented:
@TheJulianJES which quirks PR are you refering to ? 
it seems that the 2 quirk pr cited by @Caius-Bonus, ie https://github.com/home-assistant/core/pull/101709 and https://github.com/home-assistant/core/pull/102489 have been merged

AllardKatan commented:
> @TheJulianJES which quirks PR are you refering to ? it seems that the 2 quirk pr cited by @Caius-Bonus, ie #101709 and #102489 have been merged

The [#2150](https://github.com/zigpy/zha-device-handlers/pull/2150) is not merged yet, I guess that needs to be done too?

Caius-Bonus committed:
fix bitmaps and add enum for algorithm scale factor

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
improve readability if bitmapsensor

Caius-Bonus committed:
fix capitals

Caius-Bonus committed:
better setpoint response time

TheJulianJES commented on homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py:
```diff
@@ -443,3 +450,81 @@ def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+            "adaptation_run_settings": True,
+        }
+
+        super().__init__(cluster, endpoint)
```
Shouldn't this be moved to the top? If the constructor of the base class were to override/update some attributes, it should happen before we do that, not after.
So, I'd move this to the top (and I think the empty lines in the method can also be removed).

Caius-Bonus replied:
__init__ assumes REPORT_CONFIG to be populated. In this case, it wouldn't make a difference because REPORT_CONFIG is already populated, but it would be could cause bugs when adding functionality in __init__.

Caius-Bonus replied:
```
        if not hasattr(self, "_value_attribute") and self.REPORT_CONFIG:
            attr_def: ZCLAttributeDef = self.cluster.attributes_by_name[
                self.REPORT_CONFIG[0]["attr"]
            ]
            self.value_attribute = attr_def.id
```

TheJulianJES replied:
Ah, this isn't ideal. I'm no sure we still need the `value_attribute` stuff anymore.
I think we can leave it like this for now. If the `value_attribute` is eventually removed, we can also adjust this slightly.

TheJulianJES commented on homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py:
```diff
@@ -443,3 +450,81 @@ def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+
+    def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+        """Extend UserInterface."""
+
+        self.ZCL_INIT_ATTRS = {**self.ZCL_INIT_ATTRS, "viewing_direction": True}
+
+        super().__init__(cluster, endpoint)
```
Same here: move `super().__init__(cluster, endpoint)` to the top and remove empty lines.

TheJulianJES commented on homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py:
```diff
@@ -443,3 +450,81 @@ def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+
+    def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+        """Extend Diagnostic."""
+
+        self.REPORT_CONFIG = (
+            *self.REPORT_CONFIG,
+            AttrReportConfig(attr="sw_error_code", config=REPORT_CONFIG_DEFAULT),
+            AttrReportConfig(attr="motor_step_counter", config=REPORT_CONFIG_DEFAULT),
+        )
+
+        super().__init__(cluster, endpoint)
```
Same here: move `super().__init__(cluster, endpoint)` to the top and remove empty lines.

TheJulianJES commented on homeassistant/components/zha/number.py:
```diff
@@ -1053,3 +1054,74 @@ class MinHeatSetpointLimit(ZCLHeatSetpointLimitEntity):
+    _attr_native_min_value: int = 0
+    _attr_native_max_value: int = 1439
+    _attr_mode: NumberMode = NumberMode.BOX
+    _attr_native_unit_of_measurement: str = "Minutes"
```
This should probably be `UnitOfTime.MINUTES`

Caius-Bonus replied:
Done

TheJulianJES commented on homeassistant/components/zha/number.py:
```diff
@@ -1053,3 +1054,74 @@ class MinHeatSetpointLimit(ZCLHeatSetpointLimitEntity):
+    _attribute_name: str = "regulation_setpoint_offset"
+    _attr_translation_key: str = "regulation_setpoint_offset"
+    _attr_mode: NumberMode = NumberMode.BOX
+    _attr_native_unit_of_measurement: str = "°C"
```
`UnitOfTemperature.CELSIUS`

Caius-Bonus replied:
Done

TheJulianJES commented on homeassistant/components/zha/strings.json:
```diff
@@ -810,6 +831,56 @@
       "keypad_lockout": {
         "name": "Keypad lockout"
+      },
+      "exercise_day_of_week": {
+        "name": "Exercise day of the week",
+        "state": {
+          "sunday": "[%key:common::time::sunday%]",
+          "monday": "[%key:common::time::monday%]",
+          "tuesday": "[%key:common::time::tuesday%]",
+          "wednesday": "[%key:common::time::wednesday%]",
+          "thursday": "[%key:common::time::thursday%]",
+          "friday": "[%key:common::time::friday%]",
+          "saturday": "[%key:common::time::saturday%]"
+        }
+      },
+      "valve_orientation": {
+        "name": "Valve orientation",
+        "state": {
+          "horizontal": "Horizontal",
+          "vertical": "Vertical"
+        }
+      },
```
These (and more) are all translated enums states, so that assumes https://github.com/home-assistant/core/pull/109309 is merged before this, right?

TheJulianJES commented on homeassistant/components/zha/select.py:
```diff
@@ -679,3 +684,133 @@ class KeypadLockout(ZCLEnumSelectEntity):
     _attr_icon: str = "mdi:lock"
+
+
+class DanfossExerciseDayOfTheWeekEnum(types.enum8):
+    """Day of the week."""
+
+    Sunday = 0
+    Monday = 1
+    Tuesday = 2
+    Wednesday = 3
+    Thursday = 4
+    Friday = 5
+    Saturday = 6
+    Undefined = 7
```
This enum is already defined in zha-quirks, so we can just use that too.
(I know some custom entities do it differently, but I think we changed to directly using the quirks enum if possible(?))

Caius-Bonus replied:
I'll do that everywhere. It is just difficult to test right now, because my quirks PR isn't merged yet and I have to hack my code constantly to be able to test it easily on my system.

Caius-Bonus commented on homeassistant/components/zha/sensor.py:
```diff
@@ -1382,3 +1383,165 @@ class AqaraCurtainHookStateSensor(EnumSensor):
+            key for (key, elem) in self.extra_state_attributes.items() if elem
+        ]
+
+        return "something" if binary_state_attributes else "nothing"
```
This isn't shown in the UI. It functions as the translation_key and is different for every sensor.

Caius-Bonus replied:
It is indeed confusing, but it seems to be how Home Assistant works.

TheJulianJES replied:
Ah, you're right of course!

Caius-Bonus commented on homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py:
```diff
@@ -443,3 +450,81 @@ def __init__(self, cluster: zigpy.zcl.Cluster, endpoint: Endpoint) -> None:
+        )
+
+        self.ZCL_INIT_ATTRS = {
+            **self.ZCL_INIT_ATTRS,
```
I see and I would prefer to keep that style, but REPORT_CONFIG is a tuple and has to be done this way. Doing ZCL_INIT_ATTRS differently with a .update would be more confusing in my opinion.

Caius-Bonus commented:
> Can you remove the enum state translations from this PR for now? (so it doesn't "depend" on #109309) We're not yet sure how to proceed with that translation PR.

I could also make all enums lowercase in this PR, which would force HA to use the translations. It doesn't really depend on #109309. I just wasn't sure whether it would be merged first.

Caius-Bonus commented on homeassistant/components/zha/strings.json:
```diff
@@ -810,6 +831,56 @@
+      "setpoint_response_time": {
+        "name": "Setpoint response time",
+        "state": {
+          "quick_5min": "5 Minutes",
+          "quick_10min": "10 Minutes",
+          "quick_15min": "15 Minutes",
+          "quick_25min": "25 Minutes",
+          "moderate_30min": "30 Minutes",
+          "moderate_40min": "40 Minutes",
+          "moderate_50min": "50 Minutes",
+          "moderate_60min": "60 Minutes",
+          "moderate_70min": "70 Minutes",
+          "slow_80min": "80 Minutes",
```
Done.

Caius-Bonus committed:
feedback

Caius-Bonus committed:
lowercase every enum to adhere to the translation_key standard

Caius-Bonus commented:
Now it doesn't depend on https://github.com/home-assistant/core/pull/109309 at all. It uses all translations in strings.py. However I had to duplicate the enums with some changes in this PR.

TheJulianJES commented:
https://github.com/home-assistant/core/pull/86907/commits/15f217cbf4f47838da716a56d782c78ccdb1f17e was not exactly what I meant 😄 
Can you revert that (so those enums are like all others) and completely remove the enum state translations for now?

Or we have to move forward with the enum translation PR first, but the issue there is that the quirks library could in theory expand enums (that are in quirks) without HA having a test to see if the translations are updated.
I'm not quite sure yet what's allowed there and what isn't.

Caius-Bonus commented:
> [15f217c](https://github.com/home-assistant/core/commit/15f217cbf4f47838da716a56d782c78ccdb1f17e) was not exactly what I meant 😄 Can you revert that (so those enums are like all others) and completely remove the enum state translations for now?
> 
> Or we have to move forward with the enum translation PR first, but the issue there is that the quirks library could in theory expand enums (that are in quirks) without HA having a test to see if the translations are updated. I'm not quite sure yet what's allowed there and what isn't.

I understand you didn't mean that, but this makes translations work without depending on https://github.com/home-assistant/core/pull/109309. I just would have liked the translations to work for all new entities, but I don't want to make this PR wait for the enum translations to be figured out.

Caius-Bonus committed:
remove enum state translations and use enums from quirks

Caius-Bonus committed:
correctly capitalize OrientationEnum

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Mikaka27 commented:
Is there any chance to have this in 2024.5? Thanks :)

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
bump zha dependencies; this will have to be done in a separate PR, but this aids review

Caius-Bonus committed:
accidentally removed enum

Caius-Bonus committed:
Merge branch 'home-assistant:dev' into AddDanfossEntities

Caius-Bonus committed:
tests

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Hoffmann77 commented:
Any chance this could be merged soon?

Mikaka27 commented:
Is change requested still applicable to current code?

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus commented:
It looks like all the changes still add functionality and don't remove anything. It also looks like all new code is written the way it should be. 

All tests seem like they still pass and I have been running the code for over a month without any issues. 

Seems like it is good to merge.

Caius-Bonus committed:
comment

Caius-Bonus commented:
@edenhaus Could you maybe remove the change requested? It might make it look like I didn't address all reviews yet.

Also, could somebody remove the "Breaking Change" label? It is not a breaking change, it only adds functionality. The label was erroneously added, because I thought the changes in zha-quirks were considered breaking changes (even if they were considered breaking, that has nothing to do with this PR and also is already merged over a month ago).

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

puddly committed:
Migrate reporting and ZCL attribute config out of `__init__`

puddly commented:
This PR looks good to me. I've made a small PR (https://github.com/Caius-Bonus/core/pull/4) to @Caius-Bonus's branch to remove code out of `__init__`.

@TheJulianJES You left a few review comments regarding `__init__`. Would this take care of them?

TheJulianJES commented:
From a quick look, yeah. I hoped that we could get this into quirks v2 instead, as it should all be possible now, but we can can also add it here I guess.

Caius-Bonus committed:
Merge branch 'dev' into AddDanfossEntities

Caius-Bonus committed:
Merge pull request #4 from puddly/AddDanfossEntities

Migrate reporting and ZCL attribute config out of `__init__`

Caius-Bonus committed:
hvac.py shouldn't be changed in this pull request

Caius-Bonus committed:
change wording comment

Caius-Bonus committed:
I forgot I changed the size of the tuple.

Caius-Bonus commented:
@puddly Thanks, I merged your PR.

@TheJulianJES I changed the description of my PR to better reflect all changes. This PR needs to happen to at least add the BitMapSensor to HomeAssistant Core. When changing this device to Quirks V2, BitMapSensor would also need to be integrated into Quirks V2. Also, many parts of this device deviate from other devices, which means doing this in Quirks V2 would require multiple other changes to Quirks V2. 

I think it would be best to merge this PR. Then I can make a new Pull Requests to convert it to Quirks V2. This way the device works for all other users and it will be possible to change the device to Quirks V2 at a slower pace (which I think is going to take multiple PRs, because of the many deviations of this device).
